feat: Add retry implementation inspired by async-retry.

yeldiRium opened the pull request:
I took the main logic from [async-retry](https://github.com/zeit/async-retry) and rewrote it to be more in line with our style.

- I changed the default for the `randomize` parameter back to `false` to be more in line with the original settings from [retry](https://github.com/tim-kos/node-retry).
- I rewrote the code to use async/await instead of the previous `Promise.resolve(...)` approach, because async/await is much more readable in this case and reduces the complexity of error handling.

goloroden commented on test/unit/retryTests.ts:
```diff
@@ -0,0 +1,132 @@
+      await retry(
+        async (bail, num): Promise<void> => {
+          if (num === 2) {
+            bail(new Error('Wont retry'));
```
`Wont` should be `Won't`, I guess 😉 

Besides that – is there a `return` missing here?

yeldiRium commented:
As discussed I rewrote the implementation using `retryIgnoreAbort` and removed the dependency to `retry`. I also updated and expanded the documentation and removed references to `async-retry` since we have nothing to do with it anymore.
